### PR TITLE
Add configurable enable-k8s-endpoint-slice

### LINKED ISSUE #12513: Cilium does not correlate endpoints with service
Cilium does not correlate endpoins with service when endpoints are manager 'manually' (i.e. by operator). Bug was found when I was experimenting with zolando postgresql operator - operator creates service without selector and later creates endpoint to postgresql master. And it's ok for kubernetes according to [docs](https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors).

- Cilium version
Client: 1.8.1 5ce2bc7b3 2020-07-02T20:04:47+02:00 go version go1.14.4 linux/amd64
Daemon: 1.8.1 5ce2bc7b3 2020-07-02T20:04:47+02:00 go version go1.14.4 linux/amd64

- Kernel version - Linux k8s-1 5.7.0-gentoo #1 SMP PREEMPT Wed Jun 3 18:34:14 -00 2020 x86_64 Intel Core Processor (Skylake, IBRS) GenuineIntel GNU/Linux

- Orchestration system version in use:
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"archive", BuildDate:"2020-07-13T10:27:10Z", GoVersion:"go1.14.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.5", GitCommit:"e6503f8d8f769ace2f338794c914a96fc335df0f", GitTreeState:"clean", BuildDate:"2020-06-26T03:39:24Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}

[cilium-sysdump-20200713-175126.zip](https://github.com/cilium/cilium/files/4913231/cilium-sysdump-20200713-175126.zip)

**How to reproduce the issue**

1. Create test service without selector:

```
kind: Pod
apiVersion: v1
metadata:
  name: test-app
  labels:
    app: test
spec:
  containers:
    - name: test-app
      image: hashicorp/http-echo
      args:
        - "-text=test"
---
kind: Service
apiVersion: v1
metadata:
  name: test-service
spec:
  ports:
    - port: 5678 # Default port for image
```

2. Wait for service and app to start, retrieve current pod ip and create endpoint:

```
kind: Endpoints
apiVersion: v1
metadata:
  name: test-service
  namespace: default
subsets:
  - addresses:
    - ip: <POD IP>
    ports:
      - port: 5678
        protocol: TCP
```

3. Get test-service endpoints with kubectl - our endpoint will be available:

```
kubectl get endpoints test-service
NAME           ENDPOINTS           AGE
test-service   10.244.3.167:5678   8m27s
```

4. Run `cilium bpf lb list` - there will be NO ip of our service. As a result you're not able to connect to service on 5678 port.

### PR DESCRIPTION
**Pull request by Antiarchitect**
due to a bug with K8s 1.18 https://github.com/cilium/cilium/issues/12513#issuecomment-658260758

Fixes: #12513

### CODE REVIEW

**Comment by Antiarchitect** [comment]
As you can see it seems like K8s 1.18 users will suffer without this: https://github.com/kubernetes/kubernetes/pull/91637#issuecomment-683990297

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml [hunk 1]
```diff
@@ -611,3 +611,9 @@ data:
 {{- else if ( ne $crdWaitTimeout "5m" ) }}
   crd-wait-timeout: {{ $crdWaitTimeout | quote }}
 {{- end }}
+
+{{- if hasKey .Values "enableK8sEndpointSlice" }}
```
Comment (brb) [inline install/kubernetes/cilium/charts/config/templates/configmap.yaml:615]:
```suggestion
{{- if hasKey .Values "enableK8sEndpointSlice" }}
  enable-k8s-endpoint-slice: {{ .Values.enableK8sEndpointSlice | quote }}
{{- end }}  
```

**Review by joestringer — CHANGES_REQUESTED** [review]
LGTM.

One of the commits is missing signoff lines, there should be one from each author.

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml [hunk 2]
```diff
@@ -611,3 +611,11 @@ data:
...
+  enable-k8s-endpoint-slice: {{ .Values.enableK8sEndpointSlice | quote }}
+{{- end }}  
+{{- if not .Values.enableK8sEndpointSlice }}
+  enable-k8s-endpoint-slice: "false"
+{{- end }}
+{{- end }}
```
Comment (sayboras) [inline install/kubernetes/cilium/charts/config/templates/configmap.yaml:617]: This `{{- end }}` seems not required.

**Review by sayboras — CHANGES_REQUESTED** [review]
Please help to re-generate yaml manefiest and check in the files as well. You can run `make -C install/kubernetes all` to do so, this will perform basic linting as well.

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml (hunk 2 above)
Comment (aanm) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml:617]: @sayboras why not? this `{{- end }}` is referring to the `if` of line 615, right?

**Review by aanm — APPROVED** [review]
LGTM. Nothing else to point out besides what other reviewers already wrote.

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml (hunk 2 above)
Comment (sayboras) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml:617]: Is `{{- end }}` of line 617 closing if for line line 615 ? 

```
{{- if hasKey .Values "enableK8sEndpointSlice" }}
  enable-k8s-endpoint-slice: {{ .Values.enableK8sEndpointSlice | quote }}
{{- end }}  
{{- if not .Values.enableK8sEndpointSlice }}
  enable-k8s-endpoint-slice: "false"
{{- end }}
{{- end }}
```

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml [hunk 3]
```diff
@@ -611,3 +611,11 @@ data:
 {{- else if ( ne $crdWaitTimeout "5m" ) }}
   crd-wait-timeout: {{ $crdWaitTimeout | quote }}
 {{- end }}
+
+{{- if hasKey .Values "enableK8sEndpointSlice" }}
+  enable-k8s-endpoint-slice: {{ .Values.enableK8sEndpointSlice | quote }}
+{{- end }}  
+{{- if not .Values.enableK8sEndpointSlice }}
```
Comment (brb) [inline install/kubernetes/cilium/charts/config/templates/configmap.yaml]: Why do we need this case? If users want to disable it, then they should `--set config.enableK8sEndpointSlice=false`?

Comment (Antiarchitect) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml]: As it is enabled by default my initial intent was to add enable-k8s-endpoint-slice key to ConfigMap only in false case

Comment (brb) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml]: I'd keep it simple - in order to disable the feature users would be required to explicitly set `"false"`.

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml (hunk 2 above)
Comment (aanm) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml:617]: Ah I totally miss that, you are right.

File: install/kubernetes/cilium/charts/config/templates/configmap.yaml (hunk 3 above)
Comment (Antiarchitect) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml]: So what the final variant should be?

Comment (brb) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml]:
```
{{- if hasKey .Values "enableK8sEndpointSlice" }}
  enable-k8s-endpoint-slice: {{ .Values.enableK8sEndpointSlice | quote }}
{{- end }}  
```

Comment (Antiarchitect) [reply install/kubernetes/cilium/charts/config/templates/configmap.yaml]: Just did the force push

**Comment by aanm** [comment]
test-me-please

**Review by sayboras — APPROVED** [review]
LGTM, thanks :+1:

**Review by brb — APPROVED** [review]
Thanks!

**Review by christarazi — APPROVED** [review]

**Comment by christarazi** [comment]
test-me-please

**Pull request merged**